hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieAppendHandle.java

```diff
@@ -434,16 +436,20 @@ protected void processAppendResult(AppendResult result, Option<HoodieLogBlock> d
     updateWriteStatus(result, stat);
 
     if (config.isMetadataColumnStatsIndexEnabled()) {
+      HoodieSchema writerHoodieSchemaWithMetaFields = HoodieSchema.fromAvroSchema(writeSchemaWithMetaFields);
       HoodieIndexVersion indexVersion = HoodieTableMetadataUtil.existingIndexVersionOrDefault(PARTITION_NAME_COLUMN_STATS, hoodieTable.getMetaClient());
       Set<String> columnsToIndexSet = new HashSet<>(HoodieTableMetadataUtil
           .getColumnsToIndex(hoodieTable.getMetaClient().getTableConfig(),
-              config.getMetadataConfig(), Lazy.eagerly(Option.of(writeSchemaWithMetaFields)),
+              config.getMetadataConfig(), Lazy.eagerly(Option.of(writerHoodieSchemaWithMetaFields)),
               Option.of(this.recordMerger.getRecordType()), indexVersion).keySet());
-      final List<Pair<String, Schema.Field>> fieldsToIndex = columnsToIndexSet.stream()
-          .map(fieldName -> HoodieAvroUtils.getSchemaForField(writeSchemaWithMetaFields, fieldName)).collect(Collectors.toList());
+      final List<Pair<String, HoodieSchemaField>> fieldsToIndex = columnsToIndexSet.stream()
+          .map(fieldName -> {
+            Pair<String, Schema.Field> avroFieldPair = HoodieAvroUtils.getSchemaForField(writeSchemaWithMetaFields, fieldName);
+            return Pair.of(avroFieldPair.getKey(), new HoodieSchemaField(avroFieldPair.getValue()));
```

Comment:
Let's simplify this by fetching the field schema directly from `writerHoodieSchemaWithMetaFields` instead of `writeSchemaWithMetaFields`

Reply:
Done